suite/tests/client-interface/low_on_memory.dll.c

```diff
@@ -0,0 +1,142 @@
+/* **********************************************************
+ * Copyright (c) 2011-2019 Google, Inc.  All rights reserved.
+ * Copyright (c) 2008-2010 VMware, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * API regression test that registers for all supported event callbacks
+ * (except the nudge and security violation callback)
+ */
+
+#include "dr_api.h"
+#include "client_tools.h"
+#include "drmgr.h"
+#include "drwrap.h"
+
+typedef struct node_type {
+
+    int int_array[50000];
+    struct node_type *next;
+
+} node_t;
+
+static bool is_clear;
+static node_t *head;
+
+static void
+insert_new_node()
+{
+
+    node_t **node = &head;
+
+    while (*node != NULL) {
+        node = &((*node)->next);
+    }
+
+    *node = dr_global_alloc(sizeof(node_t));
+    (*node)->next = NULL;
+}
+
+static void
+wrap_pre(void *wrapcxt, OUT void **user_data)
+{
+    if (!is_clear)
+        insert_new_node();
+}
+
+static void
+module_load_event(void *drcontext, const module_data_t *mod, bool loaded)
+{
+    app_pc towrap = (app_pc)dr_get_proc_address(mod->handle, "malloc");
+    if (towrap != NULL) {
+        drwrap_wrap(towrap, wrap_pre, NULL);
+    }
+}
+
+static void
+low_on_memory_event()
+{
+    if (!is_clear) {
+        if (head == NULL)
```

Comment:
So this is assuming that either A) the app has only one thread or B) DR is synchronizing this callback for the client -- yet that's not documented.  Please add comments explaining what the assumption is (if it's B please make that part of the interface: and it prob needs another lock).

Reply:
Done